lib/common/socket/evloop.c.h

```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
+}
+
+#if defined(__linux__) && !H2O_USE_LIBUV
+static void handle_timestamp(struct st_h2o_evloop_socket_t *sock, struct msghdr *msg)
+{
+    struct scm_timestamping *ts = NULL;
+    uint64_t packet_ts = 0;
+    h2o_loop_t *loop = sock->loop;
+
+    for (struct cmsghdr *cmsg = CMSG_FIRSTHDR(msg); cmsg;
+            cmsg = CMSG_NXTHDR(msg, cmsg)) {
+
+        /* in the future we may want to harvest IP_RECVERR messages here, which could have interesting data */
+        if (cmsg->cmsg_level != SOL_SOCKET)
+            continue;
+
+        switch (cmsg->cmsg_type) {
+            case SO_TIMESTAMPNS:
+                ts = (struct scm_timestamping *)CMSG_DATA(cmsg);
+                packet_ts = timespec_to_nsec(&ts->ts[0]);
+
+                h2o_sliding_counter_start(&loop->packet_latency_nanosec_counter, packet_ts);
+                break;
+            default:
+                /* Ignore other cmsg options */
+                break;
+        }
+    }
+
+    return;
+}
+#else /* !defined(__linux__) || (defined(__linux__) && H2O_USE_LIBUV) */
+static void handle_timestamp(struct st_h2o_evloop_socket_t *sock, struct msghdr *msg)
+{
+    return;
+}
+#endif /* defined(__linux__) && !defined(H2O_USE_LIBUV) */
+
+static const char *on_read_core(struct st_h2o_evloop_socket_t *sock)
 {
     ssize_t read_so_far = 0;
+    int fd = sock->fd;
+    h2o_buffer_t **input = sock->super.ssl == NULL ? &sock->super.input : &sock->super.ssl->input.encrypted;
+    int timestamping = sock->super.timestamping;
+
+    struct msghdr msg;
 
     while (1) {
         ssize_t rret;
         h2o_iovec_t buf = h2o_buffer_try_reserve(input, 4096);
+        size_t buflen = (buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 +1);
+
         if (buf.base == NULL) {
             /* memory allocation failed */
             return h2o_socket_error_out_of_memory;
         }
-        while ((rret = read(fd, buf.base, buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 + 1)) == -1 && errno == EINTR)
+
+        /* if timestamping is enabled, lets setup everything we need to make it happen */
+        size_t controllen = CMSG_SPACE(sizeof(struct timespec));
+        char control[controllen];
+        struct iovec iov = { .iov_base = buf.base, .iov_len = buflen };
+
+        msg = (struct msghdr) { .msg_control = NULL,
+                                .msg_controllen = 0,
+                                .msg_name = NULL,
+                                .msg_namelen = 0,
+                                .msg_iov = &iov,
+                                .msg_iovlen = 1 };
+
+        if (timestamping) {
+            memset(control, 0, controllen);
+            msg.msg_control = &control;
+            msg.msg_controllen = controllen;
+        }
+
+        while ((rret=recvmsg(fd, &msg, 0)) == -1 && errno == EINTR)
             ;
+
+        if (rret > 0) {
+            if (sock->super.needs_timestamp) {
+                handle_timestamp(sock, &msg);
+                sock->super.needs_timestamp = 0;
```

Comment:
Is the intent here to collect timestamp once per each connection? If so, could you clarify what is the benefit for discarding the variables that have been collected mid-connection, please?